maps/mountain_fortress_v3/terrain.lua

```diff
@@ -146,6 +146,78 @@ local function shuffle(tbl)
     return tbl
 end
 
+--- Returns a function that will generate the next value each time it is called.
+--- It is configured by the list of such elements {value = '...', tags = {}, weight = 100}.
+--- It ensures that consecutive values do not have the same tags (if such variants exist).
+--- The probability of generating the next value is calculated based on the weight parameter.
+local function generate_with_tags_and_weights(variants)
+    local prev_tags = {}
+
+    for _, variant in pairs(variants) do
+        -- Inner weight set to 0 every time we generate the corresponding value.
+        -- Each iteration it grows up to its originally weight.
+        variant.inner_weight = variant.weight
+    end
+
+    local function intersect(arr1, arr2)
+        local result = {}
+        for _, i1 in pairs(arr1) do
+            for _, i2 in pairs(arr2) do
+                if i1 == i2 then
+                    table.insert(result, i1)
+                end
+            end
+        end
+        return result
+    end
+
+    function next_value_provider()
```

Comment:
Why aren't we utilizing the local variable here? The workflow is not happy :}

Reply:
**next_value_provider** is actually a variable, it is created within the **generate_with_tags_and_weights** function.
What exactly is wrong? I seriously don't get it :) We create inner function, return it and call from the "client" function (init_terrain).
**next_value_provider** returns new value each time, but **prev_tags** also keep tags of previous zone. 
If you mean we should pass prev tags throught the parameter... yeah, but in this case we are moving part of our logics outside, the incapsulation is breaking :(
I didn't find classes in lua, so have to do things like this.

Reply:
Sorry for my english :sweat_smile:

Reply:
Oh, if you mean the **intersect** function, it's an utility, so I couldn't decide where to put it, so I nested it

Reply:
I guess I got it. You mean the "local" one. Sorry, my first time with lua :)